src/main/scala/is/hail/utils/richUtils/RichIndexedRowMatrix.scala

```diff
@@ -1,50 +1,115 @@
     val excessRows = (rows % blockSize).toInt
     val excessCols = (cols % blockSize).toInt
+    val fullColPartitions = if (excessCols == 0) colPartitions else colPartitions - 1
 
     val blocks = indexedRowMatrix.rows.flatMap { ir =>
-      val i = ir.index / blockSize
-      val ii = ir.index % blockSize
-
-      ir.vector.toArray
-        .grouped(blockSize)
-        .zipWithIndex
-        .map { case (values, blockColumn) =>
-          ((i.toInt, blockColumn), (ii.toInt, values))
-        }
-    }.groupByKey(GridPartitioner(rows, cols, blockSize)).mapValuesWithKey {
-      case ((i, j), it) =>
+      val i = (ir.index / blockSize).toInt
+      val ii = (ir.index % blockSize).toInt
+      val a = ir.vector.toArray
+      val grouped = new Array[((Int, Int), (Int, Int, Int, Array[Double]))](colPartitions)
+
+      var j = 0
+      while (j < fullColPartitions) {
+        val group = new Array[Double](blockSize)
+        grouped(j) = ((i, j), (i, j, ii, group))
+        System.arraycopy(a, j * blockSize, group, 0, blockSize)
+        j += 1
+      }
+      if (excessCols > 0) {
+        val group = new Array[Double](excessCols)
+        grouped(j) = ((i, j), (i, j, ii, group))
+        System.arraycopy(a, j * blockSize, group, 0, excessCols)
+      }
+
+      grouped.iterator
+    }.aggregateByKey(null: Array[Double], partitioner)(
+      seqOp(truncatedBlockRow, truncatedBlockCol, excessRows, excessCols, blockSize), combOp)
+      .mapValuesWithKey { case ((i, j), a) =>
         val rowsInBlock: Int = if (i == truncatedBlockRow) excessRows else blockSize
         val colsInBlock: Int = if (j == truncatedBlockCol) excessCols else blockSize
-
-        val a = new Array[Double](rowsInBlock * colsInBlock)
-        it.foreach{ case (ii, values) =>
-          var jj = 0
-          while (jj < values.length) {
-            a(jj * rowsInBlock + ii) = values(jj)
-            jj += 1
-          }
-        }
         new BDM[Double](rowsInBlock, colsInBlock, a)
     }
-    new BlockMatrix(blocks, blockSize, rows, cols)
+
+    new BlockMatrix(new EmptyPartitionIsAZeroMatrixRDD(blocks), blockSize, rows, cols)
   }
+
+}
+
+private class EmptyPartitionIsAZeroMatrixRDD(blocks: RDD[((Int, Int), BDM[Double])])
+    extends RDD[((Int, Int), BDM[Double])](blocks.sparkContext, Seq[Dependency[_]](new OneToOneDependency(blocks))) {
+  assert(!blocks.partitioner.isEmpty)
+  assert(blocks.partitioner.get.isInstanceOf[GridPartitioner])
+
+  val gp = blocks.partitioner.get.asInstanceOf[GridPartitioner]
+
+  def compute(split: Partition, context: TaskContext): Iterator[((Int, Int), BDM[Double])] = {
+    val it = blocks.iterator(split, context)
+    if (it.hasNext)
+      it
+    else
+      Iterator.single(gp.blockCoordinates(split.index) -> BDM.zeros[Double](gp.blockSize, gp.blockSize))
+  }
+
+  protected def getPartitions: Array[Partition] =
+    blocks.partitions
+
+  @transient override val partitioner: Option[Partitioner] =
```

Comment:
gp is still getting serialized because it's not transient up at the definition. This can be a def, and the val gp above should be transient.